google/api_core/rest_helpers.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for rest transports."""
+
+import itertools
+
+
+def flatten_query_params(obj, key_path=[]):
+    """Flatten a nested dict into a list of (name,value) tuples.
+
+    The result is suitable for setting query params on an http request.
+
+    .. code-block:: python
+
+        >>> obj = {'a':
+        ...         {'b':
+        ...           {'c': ['x', 'y', 'z']} },
+        ...      'd': 'uvw', }
+        >>> flatten_query_params(obj)
+        [('a.b.c', 'x'), ('a.b.c', 'y'), ('a.b.c', 'z'), ('d', 'uvw')]
+
+    Args:
+      obj: a nested dictionary (from json)
+      key_path: a list of name segments, representing levels above this obj.
+
+    Returns: a list of tuples, with each tuple having a (possibly) multi-part name
+      and a scalar value.
+    """
+
+    if obj is None:
+        return []
+    if isinstance(obj, dict):
+        return _flatten_dict(obj, key_path=key_path)
+    if isinstance(obj, list):
+        return _flatten_list(obj, key_path=key_path)
+    return _flatten_value(obj, key_path=key_path)
+
+
+def _is_value(obj):
+    if obj is None:
+        return False
+    return not (isinstance(obj, list) or isinstance(obj, dict))
+
+
+def _flatten_value(obj, key_path=[]):
+    if not key_path:
+        # There must be a key.
+        return []
```

Comment:
This never happens:  the caller always passes a real path:
```suggestion
```

Reply:
_flatten_value can be called from flatten_query_params, which should be called with a dict, but in theory could be called as `flatten_query_params("abc")`.

Reply:
What would the caller's intent be there?  This is another case where having the function return an empty list silently is surprising:  `flatten_query_params("abc")` should probably raise an exception.

Reply:
That's fair.  I factored out a top-level function to check the top-level type, from the helper function to check inner types.  The top level raises an exception if the input is not a dict or None.